Source/player.cpp

```diff
@@ -1937,34 +1937,35 @@ bool Player::HasItem(int item, int *idx) const
 
 void Player::RemoveInvItem(int iv, bool calcScrolls)
 {
-	iv++;
-
 	// Iterate through invGrid and remove every reference to item
-	for (int8_t &itemId : InvGrid) {
-		if (itemId == iv || itemId == -iv) {
-			itemId = 0;
+	for (int8_t &itemIndex : InvGrid) {
+		if (abs(itemIndex) - 1 == iv) {
+			itemIndex = 0;
 		}
 	}
 
-	iv--;
+	InvList[iv]._itype = ItemType::None;
+
 	_pNumInv--;
 
 	// If the item at the end of inventory array isn't the one we removed, we need to swap its position in the array with the removed item
 	if (_pNumInv > 0 && _pNumInv != iv) {
 		InvList[iv] = InvList[_pNumInv];
+		InvList[_pNumInv]._itype = ItemType::None;

```

Comment:
Looks good. I'd add a `memset(&InvList[_pNumInv], 0, sizeof(ItemStruct))` prior to this line, just to clear more garbage data from memory. And do the same on the above `._itype = ItemType::None;` line.

Reply:
DevilutionX is C++: `InvList[_pNumInv] = {};` ;)

Reply:
Going to opt out of zeroing just to keep thing simple for now.